### PR TITLE
feature(new integration): Canny Destination Onboarding

### PR DESCRIPTION
**Pull request by rohithkaza**
## Description of the change

> Onboarding Canny Destination

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/canny/transform.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+      throw new CustomError("Message type not supported", 400);
+  }
+  return response;
+};
+
+const process = async event => {
+  return await processEvent(event.message, event.destination);
+};
+
+module.exports = { process };
```
Comment (ItsSudip) [inline v0/destinations/canny/transform.js:174]: Add `processRouterDest` function.

File: v0/destinations/canny/transform.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+};
+
+const trackResponseBuilder = async (message, { Config }) => {
+  const { apiKey } = Config;
+  const eventsToEvents = Config.eventsToEvents;
+  const eventsMap = getHashFromArrayWithDuplicate(eventsToEvents);
+
+  if (!apiKey) {
+    throw new CustomError("API Key is not present. Aborting message.", 400);
+  }
```
Comment (ItsSudip) [inline v0/destinations/canny/transform.js:116]: This is repeating put them inside process function.

File: v0/destinations/canny/transform.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  if (!eventsMap[event]) {
+    throw new CustomError(
+      `Event name (${event}) is not present in the mapping`,
+      400
+    );
+  }
+
+  const returnArray = [];
+  const eventArray = Object.keys(eventsMap);
+  for (let x = 0; x < eventArray.length; x++) {
```
Comment (ItsSudip) [inline v0/destinations/canny/transform.js:132]: Try using for .... of.

File: v0/destinations/canny/transform.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    //   Object.keys(eventsMap).forEach(async key => {
```
Comment (ItsSudip) [inline v0/destinations/canny/transform.js:133]:
```suggestion
```

File: v0/destinations/canny/transform.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    if (eventArray[x] === event) {
+      const eventsMapArray = eventsMap[event];
+      for (const element of eventsMapArray) {
+        //   eventsMap[event].forEach(async val => {
```
Comment (ItsSudip) [inline v0/destinations/canny/transform.js:137]:
```suggestion
```

File: v0/destinations/canny/transform.js (hunk 1 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Added

File: v0/destinations/canny/transform.js (hunk 2 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js (hunk 3 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js (hunk 4 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Removed

File: v0/destinations/canny/transform.js (hunk 5 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Removed

File: v0/destinations/canny/data/createPostConfig.json [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        "properties.categoryID",
+        "properties.categoryId",
+        "properties.category.id"
+      ],
+    "sourceFromGenericMap": false,
+    "required": false
+  },
+  {
+    "destKey": "customFields",
+    "sourceKeys": "properties.customFields",
```
Comment (ItsSudip) [inline v0/destinations/canny/data/createPostConfig.json:34]: This map is not correct. We should put everything from properties after removing boardID, byID, categoryID,details, eta, etaPublic, title, imageURLs.

File: v0/destinations/canny/util.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Function to retrieve userId from canny using axios
+ * @param apiKey
+ * @param message
+ * @returns canny userId
+ */
+const retrieveUserId = async (apiKey, message) => {
+  try {
+    if (message?.context?.externalId?.type === "cannyUserId") {
```
Comment (ItsSudip) [inline v0/destinations/canny/util.js:13]: Use this function instead `getDestinationExternalID`.

File: v0/destinations/canny/data/createPostConfig.json (hunk 6 above)
Comment (rohithkaza) [reply v0/destinations/canny/data/createPostConfig.json:34]: User should create each and every customField on Canny's Dashboard and then should send them here, orelse the call will not happen. Won't it cause a problem?

File: v0/destinations/canny/util.js (hunk 7 above)
Comment (rohithkaza) [reply v0/destinations/canny/util.js]: Done

File: v0/destinations/canny/data/createPostConfig.json (hunk 6 above)
Comment (ItsSudip) [reply v0/destinations/canny/data/createPostConfig.json:34]: We will share these details in the doc. But let's not take it as `customFields` key.

File: v0/destinations/canny/transform.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  const eventArray = Object.keys(eventsMap);
+  for (const key of eventArray) {
+    if (key === event) {
+      const eventsMapArray = eventsMap[event];
+      for (const element of eventsMapArray) {
+        const response = await getTrackResponse(apiKey, message, element);
+        returnArray.push(response);
+      }
+    }
+  }
```
Comment (utsabc) [inline v0/destinations/canny/transform.js:134]:
```suggestion
  for (const configuredSourceEvent of configuredSourceEvents) {
    if (configuredSourceEvent === message.event) {
      const destinationEvents = configuredEventsMap[message.event];
      for (const destinationEvent of destinationEvents) {
        const response = await getTrackResponse(apiKey, message, destinationEvent);
        responseArray.push(response);
      }
    }
  }
```

File: v0/destinations/canny/transform.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+    if (!payload.boardID) {
+      throw new CustomError("BoardID is not present. Aborting message.", 400);
+    }
+    if (!payload.title) {
+      throw new CustomError("Title is not present. Aborting message.", 400);
+    }
+    if (!payload.details) {
+      throw new CustomError("Details is not present. Aborting message.", 400);
+    }
```
Comment (utsabc) [inline v0/destinations/canny/transform.js:96]: Put these in a small validation fxn

File: v0/destinations/canny/transform.js [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  if (!payload.name) {
+    throw new CustomError("Name is not present. Aborting message.", 400);
+  }
+
+  const { endpoint } = ConfigCategory.IDENTIFY;
+
+  return responseBuilder(payload, apiKey, endpoint, contentType, responseBody);
+};
+
+const getTrackResponse = async (apiKey, message, val) => {
```
Comment (utsabc) [inline v0/destinations/canny/transform.js:58]:
```suggestion
const getTrackResponse = async (apiKey, message, operationType) => {
```

File: v0/destinations/canny/transform.js [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+      throw new CustomError("Details is not present. Aborting message.", 400);
+    }
+
+    payload.apiKey = apiKey;
+    payload.authorID = await retrieveUserId(apiKey, message);
+
+    endpoint = ConfigCategory.CREATE_POST.endpoint;
+  }
+
+  return responseBuilder(payload, apiKey, endpoint, contentType, responseBody);
```
Comment (utsabc) [inline v0/destinations/canny/transform.js:104]:
```suggestion
const responseConfgs = {payload, apiKey, endpoint, contentType, responseBody};
  return responseBuilder(responseConfgs);
```

File: v0/destinations/canny/transform.js [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+  if (!payload.userID) {
+    throw new CustomError("UserId is not present. Aborting message.", 400);
+  }
+  if (!payload.name) {
+    throw new CustomError("Name is not present. Aborting message.", 400);
+  }
+
+  const { endpoint } = ConfigCategory.IDENTIFY;
+
+  return responseBuilder(payload, apiKey, endpoint, contentType, responseBody);
```
Comment (utsabc) [inline v0/destinations/canny/transform.js:55]: Same as track

**Review by utsabc — CHANGES_REQUESTED** [review]

File: v0/destinations/canny/transform.js [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+    endpoint = ConfigCategory.CREATE_POST.endpoint;
+  }
+
+  return responseBuilder(payload, apiKey, endpoint, contentType, responseBody);
+};
+
+const trackResponseBuilder = async (message, { Config }) => {
+  const { apiKey } = Config;
+  const eventsToEvents = Config.eventsToEvents;
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+  if (!event) {
+    throw new CustomError("Event name is required", 400);
+  }
+
+  if (!eventsMap[event]) {
+    throw new CustomError(
+      `Event name (${event}) is not present in the mapping`,
+      400
+    );
+  }
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+
+  const returnArray = [];
+  const eventArray = Object.keys(eventsMap);
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+    endpoint = ConfigCategory.CREATE_POST.endpoint;
+  }
+
+  return responseBuilder(payload, apiKey, endpoint, contentType, responseBody);
+};
+
+const trackResponseBuilder = async (message, { Config }) => {
+  const { apiKey } = Config;
+  const eventsToEvents = Config.eventsToEvents;
+  const eventsMap = getHashFromArrayWithDuplicate(eventsToEvents);
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js (hunk 8 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js (hunk 9 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js (hunk 10 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js (hunk 11 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Done

File: v0/destinations/canny/transform.js (hunk 12 above)
Comment (rohithkaza) [reply v0/destinations/canny/transform.js]: Updated

**Review by utsabc — DISMISSED** [review]

**Review by ItsSudip — APPROVED** [review]

**Review by ItsSudip — DISMISSED** [review]

**Review by utsabc — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**